### PR TITLE
consensus: Simplify ConnectTrace

### PR DESCRIPTION
**Pull request by jnewbery**
ConnectTrace is now only used to track blocks that were connected during
an ActivateBestChainStep call. Simplify it to a typedef and fix
comments.

Rather than trying to remove ConnectTrace, which would change the order that notifications are fired in the case of a re-org (https://github.com/bitcoin/bitcoin/pull/17562#issuecomment-615251297), just simplify it and fix comments.

### CODE REVIEW

File: src/validation.cpp [hunk 1]
```diff
@@ -2512,45 +2512,20 @@ static int64_t nTimeFlush = 0;
 static int64_t nTimeChainState = 0;
 static int64_t nTimePostConnect = 0;
 
+//! A single block and block index pointer.
 struct PerBlockConnectTrace {
     CBlockIndex* pindex = nullptr;
     std::shared_ptr<const CBlock> pblock;
-    PerBlockConnectTrace() {}
+    PerBlockConnectTrace(CBlockIndex* index, std::shared_ptr<const CBlock> block) : pindex(index), pblock(block) {}
```
Comment (MarcoFalke) [inline src/validation.cpp:2519]:
```suggestion
    PerBlockConnectTrace(CBlockIndex* index, const std::shared_ptr<const CBlock>& block) : pindex(index), pblock(block) {}
```

style nit to avoid shared_ptr increment/decrement

File: src/validation.cpp [hunk 2]
```diff
@@ -2512,45 +2512,20 @@ static int64_t nTimeFlush = 0;
...
-        // blocks are added after all the conflicted transactions have
-        // been filled in. Thus, the last entry should always be an empty
-        // one waiting for the transactions from the next block. We pop
-        // the last entry here to make sure the list we return is sane.
-        assert(!blocksConnected.back().pindex);
-        blocksConnected.pop_back();
-        return blocksConnected;
-    }
-};
+typedef std::vector<PerBlockConnectTrace> ConnectTrace;
```
Comment (MarcoFalke) [inline src/validation.cpp:2528]: Why is this needed?

File: src/validation.cpp [hunk 3]
```diff
@@ -2868,7 +2843,7 @@ bool CChainState::ActivateBestChain(BlockValidationState &state, const CChainPar
                 }
                 pindexNewTip = m_chain.Tip();
 
-                for (const PerBlockConnectTrace& trace : connectTrace.GetBlocksConnected()) {
+                for (auto& trace : connectTrace) {
```
Comment (MarcoFalke) [inline src/validation.cpp:2846]: Why is this change to auto? `auto` shouldn't be used in consensus code unless there is a reason

File: src/validation.h [hunk 4]
```diff
@@ -409,7 +409,9 @@ enum DisconnectResult
     DISCONNECT_FAILED   // Something else went wrong.
 };
 
-class ConnectTrace;
+// forward declaration for ActivateBestChainStep and ConnectTip
+struct PerBlockConnectTrace;
+typedef std::vector<PerBlockConnectTrace> ConnectTrace;
```
Comment (MarcoFalke) [inline src/validation.h:414]:
```suggestion
using ConnectTrace = std::vector<PerBlockConnectTrace>;
```

style nit for C++11

**Review by maflcko — COMMENTED** [review]
Concept ACK

File: src/validation.cpp [hunk 5]
```diff
@@ -2512,45 +2512,20 @@ static int64_t nTimeFlush = 0;
...
-        blocksConnected.back().pblock = std::move(pblock);
-        blocksConnected.emplace_back();
-    }
-
-    std::vector<PerBlockConnectTrace>& GetBlocksConnected() {
-        // We always keep one extra block at the end of our list because
-        // blocks are added after all the conflicted transactions have
-        // been filled in. Thus, the last entry should always be an empty
-        // one waiting for the transactions from the next block. We pop
-        // the last entry here to make sure the list we return is sane.
```
Comment (MarcoFalke) [inline src/validation.cpp:2548]: Note to other reviewers: This comment obviously no longer applies after 312d27b

File: src/validation.cpp (hunk 1 above)
Comment (jnewbery) [reply src/validation.cpp:2519]: Thanks for this comment! It sent me down a bit of a rabbit hole about how to pass smart pointers. Here's what I've learned:

I think your suggestion to change the ctor to take a const lvalue reference makes a copy and therefore increases the ref count (the `const` is the callee saying to the caller "I won't steal this reference", and so if the callee wants to keep ownership, it needs to make a copy). We can avoid that copy in three ways:

- pass by non-const lvalue reference (`std::shared_ptr<const CBlock>&`) which is the callee saying "I may or may not steal this reference" and call the constructor with an lvalue (ie `thisBlock` below).
- pass by value (`std::shared_ptr<const CBlock>`), and have the caller cast to rvalue reference in the call (ie `std::move(thisBlock)` below). The `std::move` is the caller saying "I no longer want ownership of this reference. In this case, the caller can still keep shared ownership by passing an lvalue (`thisBlock`), which would increase the ref count.
- pass by rvalue (`std::shared_ptr<const CBlock>&&`) which is the callee saying "I will steal ownership". The caller has to call using an rvalue reference (`std::move(thisBlock)`). Calling by lvalue would be invalid.

I think it's best to keep this as pass-by-value and have the caller pass it with an rvalue reference.

Also, the initializer list needs to use a std::move to prevent the ref increment/decrement, which I'll change.

Some references:
- summary of ways to pass smart pointers as arguments (it's talking about unique_ptr, but the same concepts are true for shared_ptrs): https://stackoverflow.com/questions/8114276/how-do-i-pass-a-unique-ptr-argument-to-a-constructor-or-a-function
- CPP Core Guidelines on passing shared pointers: http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-sharedptrparam-owner
- Herb Sutter on smart pointer parameters: https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/
- Scott Meyers on passing move types by value/rvalue reference: http://scottmeyers.blogspot.com/2014/07/should-move-only-types-ever-be-passed.html
- Demonstration code of the different ways to pass a shared_ptr and the effect on ref counts: https://gist.github.com/jnewbery/10ee24a5ce783932accddbe4347f110e

File: src/validation.cpp (hunk 2 above)
Comment (jnewbery) [reply src/validation.cpp:2528]: It's not needed. I just wanted to keep it here so the comment made sense and I didn't have to move it to the header file.

File: src/validation.cpp (hunk 3 above)
Comment (jnewbery) [reply src/validation.cpp:2846]: Sounds sensible. I'll revert.

File: src/validation.h (hunk 4 above)
Comment (jnewbery) [reply src/validation.h]: I don't think it makes a difference for aliases which are not templates (and I can't see this in the style guide), but changed anyway.

**Comment by jnewbery** [comment]
Thanks for the review @MarcoFalke . I've addressed your comments.

File: src/validation.cpp (hunk 2 above)
Comment (MarcoFalke) [reply src/validation.cpp]: Didn't see the new comment, but I think you can put the new comment to the single place where connecttrace is instantiated. No strong opinion, though.

File: src/validation.h (hunk 4 above)
Comment (jonatack) [reply src/validation.h]: > I don't think it makes a difference for aliases which are not templates (and I can't see this in the style guide), but changed anyway.

This is my understanding as well after looking this up, along with `typedef` being the earlier, traditional syntax of the two. Perhaps one advantage `typedef` may have over `using` -- this is personal observation only and not for bikeshedding purposes :) -- is that it is used only for declaring type aliases.

Comment (jonatack) [reply src/validation.h]: Git grepping instances of type aliases seems easier with`typedef` as well.

**Comment by jonatack** [comment]
Concept ACK

File: src/validationinterface.h [hunk 6]
```diff
@@ -132,14 +132,19 @@ class CValidationInterface {
     virtual void TransactionRemovedFromMempool(const CTransactionRef &ptx) {}
     /**
      * Notifies listeners of a block being connected.
-     * Provides a vector of transactions evicted from the mempool as a result.
+     *
+     * The mempool is not guaranteed to be in a consistent state when this
+     * notification is fired. Listeners should use UpdatedBlockTip for that.
```
Comment (robot-dreams) [inline src/validationinterface.h:137]: Beginner question: what do you mean by "consistent state" here?

i.e. what could go wrong with the mempool after some, but not all, of the `ActivateBestChainStep` involved in `ActivateBestChain`?

**Review by robot-dreams — COMMENTED** [review]
Concept ACK, thanks for doing this cleanup!

Unit tests passed locally

(Aside: What's the desired benefit of reviewers doing local testing, given the nice CI system? Is it to get more test runs on different system configurations?)

File: src/validation.cpp [hunk 7]
```diff
@@ -2610,7 +2585,7 @@ bool CChainState::ConnectTip(BlockValidationState& state, const CChainParams& ch
     LogPrint(BCLog::BENCH, "  - Connect postprocess: %.2fms [%.2fs (%.2fms/blk)]\n", (nTime6 - nTime5) * MILLI, nTimePostConnect * MICRO, nTimePostConnect * MILLI / nBlocksTotal);
     LogPrint(BCLog::BENCH, "- Connect block: %.2fms [%.2fs (%.2fms/blk)]\n", (nTime6 - nTime1) * MILLI, nTimeTotal * MICRO, nTimeTotal * MILLI / nBlocksTotal);
 
-    connectTrace.BlockConnected(pindexNew, std::move(pthisBlock));
+    connectTrace.emplace_back(pindexNew, pthisBlock);
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/validation.cpp:2588]: Yes, you're right! The caller should cast to rvalue reference by calling `std::move()` (that was my intention in https://github.com/bitcoin/bitcoin/pull/18685#discussion_r410717136).

Thanks for catching this.

File: src/validationinterface.h (hunk 6 above)
Comment (jnewbery) [reply src/validationinterface.h:137]: cs_main and mempool.cs are both still being held when the `BlockConnected` signal is called, and `UpdateMempoolForReorg()` may have been called after blocks were connected but before the `BlockConnected` signal is called. See the comment above for the `TransactionAddedToMempool` signal.

The `UpdatedBlockTip` signal is called just before releasing `cs_main`, when we know that the mempool is consistent with the chain tip.

**Comment by jnewbery** [comment]
I've addressed the comment here https://github.com/bitcoin/bitcoin/pull/18685#discussion_r479612111 and rebased on master (there wasn't a conflict but this was a few thousand commits behind master and it's a small change).

File: src/validation.cpp [hunk 8]
```diff
@@ -2541,45 +2541,20 @@ static int64_t nTimeFlush = 0;
...
 /**
- * Used to track blocks whose transactions were applied to the UTXO state as a
- * part of a single ActivateBestChainStep call.
- *
- * This class is single-use, once you call GetBlocksConnected() you have to throw
- * it away and make a new one.
+ * Used to track blocks that were connected as part of a single
+ * ActivateBestChainStep call. The BlockConnected validationinterface signal is
+ * fired after a single ABCS call for all blocks that were connected in that
+ * ABCS call.
```
Comment (robot-dreams) [inline src/validation.cpp:2555]: Nit (feel free to ignore): How would you feel about rewording this to make it really obvious that there's one BlockConnected signal per block? For example:

    Used to track blocks that were connected as part of a single
    ActivateBestChainStep call. After the ABCS call, a separate BlockConnected
    validationinterface signal is fired for each block that was connected
    in the ABCS call.

**Review by robot-dreams — COMMENTED** [review]
ACK d956ac72f11c0888c6d719d257aa050a37d12ff2

File: src/validationinterface.h (hunk 6 above)
Comment (jnewbery) [reply src/validationinterface.h:138]: I agree with your first point. For the second point, if we consider the client to be a passive listener (ie the only input it's getting is from the validation interface), then it shouldn't/can't acquire those locks. It should wait for a `UpdatedBlockTip` to be sure that the state is consistent.

File: src/validation.cpp (hunk 8 above)
Comment (jnewbery) [reply src/validation.cpp:2555]: Sure. That seems better. Updated.

**Review by ajtowns — COMMENTED** [review]
utACK 4faaaa1d5935e1296c624a8d9d414989ec2fede6

File: src/validation.cpp [hunk 9]
```diff
@@ -2541,45 +2541,20 @@ static int64_t nTimeFlush = 0;
...
-        // blocks are added after all the conflicted transactions have
-        // been filled in. Thus, the last entry should always be an empty
-        // one waiting for the transactions from the next block. We pop
-        // the last entry here to make sure the list we return is sane.
-        assert(!blocksConnected.back().pindex);
-        blocksConnected.pop_back();
-        return blocksConnected;
-    }
-};
+using ConnectTrace = std::vector<PerBlockConnectTrace>;
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/validation.cpp:2557]: Done. I've moved everything into the class declaration in the header.

File: src/validation.cpp [hunk 10]
```diff
@@ -2541,45 +2541,20 @@ static int64_t nTimeFlush = 0;
 static int64_t nTimeChainState = 0;
 static int64_t nTimePostConnect = 0;
 
+//! A single block and block index pointer.
 struct PerBlockConnectTrace {
     CBlockIndex* pindex = nullptr;
     std::shared_ptr<const CBlock> pblock;
-    PerBlockConnectTrace() {}
+    PerBlockConnectTrace(CBlockIndex* index, std::shared_ptr<const CBlock> block) : pindex(index), pblock(std::move(block)) {}
```
[the remark this answers is not in the document]
Comment (jnewbery) [reply src/validation.cpp:2548]: Done.

**Comment by jnewbery** [comment]
Thanks for the review, @ajtowns! I've addressed your comments and rebased on master (since this was a couple of thousand commits behind the current master).

**Comment by jnewbery** [comment]
@ajtowns do you mind looking at this again? I've addressed your review comments.

**Comment by jnewbery** [comment]
Closing due to lack of reviewer interest.

**Comment by Rspigler** [comment]
Mark up for grabs?

**Comment by jnewbery** [comment]
> Mark up for grabs?

@Rspigler feel free to pick this up if you're interested. I'm happy to review, but this didn't get much reviewer interest last time round.
